### PR TITLE
Cyber observables

### PR DESCRIPTION
**Pull request by rpiazza**

### CODE REVIEW

File: stix2/observables.py [hunk 1]
```diff
@@ -184,11 +281,69 @@ class Mutex(_Observable):
...
 
 
+class HTTPRequestExt(_STIXBase):
+    _properties = {
+        'request_method': StringProperty(),
+        'request_value': StringProperty(),
+        'request_version': StringProperty(),
+        'request_header': DictionaryProperty(),
+        'message_body_length': IntegerProperty(),
+        'message_body_data_ref': ObjectReferenceProperty(),
```
Comment (clenk) [inline stix2/observables.py]: _check_property() in _Observable should check object references inside extensions. We'll need to add a test for this.

File: stix2/observables.py [hunk 2]
```diff
@@ -255,14 +453,23 @@ class URL(_Observable):
     }
 
 
+class UNIXAccountExt(_STIXBase):
+    _properties = {
+        'gid': IntegerProperty(),
+        'groups': ListProperty(StringProperty),
+        'home_dir': StringProperty(),
+        'shell': StringProperty(),
+    }
```
Comment (clenk) [inline stix2/observables.py]: Should have an object constraint - all 4 properties are optional but at least one of them must be present.

File: stix2/observables.py [hunk 3]
```diff
@@ -213,11 +368,54 @@ def _check_object_constaints(self):
...
+        'priority': StringProperty(),
+        'owner_sid': StringProperty(),
+        'window_title': StringProperty(),
+        'startup_info': DictionaryProperty(),
+    }
+
+
+class WindowsServiceExt(_STIXBase):
+    _properties = {
+        'service_name': StringProperty(),
```
Comment (clenk) [inline stix2/observables.py]: This property is required.

File: stix2/observables.py [hunk 4]
```diff
@@ -213,11 +368,54 @@ def _check_object_constaints(self):
...
 
+class WindowsProcessExt(_STIXBase):
+    _properties = {
+        'aslr_enabled': BooleanProperty(),
+        'dep_enabled': BooleanProperty(),
+        'priority': StringProperty(),
+        'owner_sid': StringProperty(),
+        'window_title': StringProperty(),
+        'startup_info': DictionaryProperty(),
+    }
```
Comment (clenk) [inline stix2/observables.py]: Should have an object constraint - all properties are optional but at least one of them must be present.

File: stix2/observables.py [hunk 5]
```diff
@@ -184,11 +281,69 @@ class Mutex(_Observable):
...
+class ICMPExt(_STIXBase):
+    _properties = {
+        'icmp_type_hex': HexProperty(),
+        'icmp_code_hex': HexProperty(),
+    }
+
+
+class SocketExt(_STIXBase):
+    _properties = {
+        'address_family': EnumProperty([
```
Comment (clenk) [inline stix2/observables.py]: This property is required.

File: stix2/observables.py [hunk 6]
```diff
@@ -184,11 +281,69 @@ class Mutex(_Observable):
...
+        'message_body_length': IntegerProperty(),
+        'message_body_data_ref': ObjectReferenceProperty(),
+    }
+
+
+class ICMPExt(_STIXBase):
+    _properties = {
+        'icmp_type_hex': HexProperty(),
+        'icmp_code_hex': HexProperty(),
+    }
```
Comment (clenk) [inline stix2/observables.py]: Both properties are required.

File: stix2/observables.py [hunk 7]
```diff
@@ -184,11 +281,69 @@ class Mutex(_Observable):
     }
 
 
+class HTTPRequestExt(_STIXBase):
+    _properties = {
+        'request_method': StringProperty(),
+        'request_value': StringProperty(),
```
Comment (clenk) [inline stix2/observables.py]: `request_method` and `request_value` are required.

**Comment by clenk** [comment]
The extension types that don't have any required properties should have an object constraint - all properties are optional but at least one of them must be present.

- NTFSExt
- PDFExt
- RasterImageExt
- TCPExt
- WindowsProcessExt
- UNIXAccountExt

File: stix2/observables.py (hunk 3 above)
Comment (rpiazza) [inline stix2/observables.py]: Done

File: stix2/observables.py (hunk 5 above)
Comment (rpiazza) [inline stix2/observables.py]: Done

File: stix2/observables.py (hunk 6 above)
Comment (rpiazza) [inline stix2/observables.py]: Done

File: stix2/observables.py (hunk 7 above)
Comment (rpiazza) [inline stix2/observables.py]: Done

**Comment by clenk** [comment]
Thanks, @rpiazza! We'll add the object constraints later.

**Pull request merged**